sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala

```diff
@@ -132,28 +131,21 @@ object InjectRuntimeFilter extends Rule[LogicalPlan] with PredicateHelper with J
       REGEXP_EXTRACT_FAMILY, REGEXP_REPLACE)
   }
 
-  private def canFilterLeft(joinType: JoinType): Boolean = joinType match {
-    case Inner | RightOuter => true
-    case _ => false
-  }
-
-  private def canFilterRight(joinType: JoinType): Boolean = joinType match {
-    case Inner | LeftOuter => true
-    case _ => false
-  }
-
   private def isProbablyShuffleJoin(left: LogicalPlan,
       right: LogicalPlan, hint: JoinHint): Boolean = {
     !hintToBroadcastLeft(hint) && !hintToBroadcastRight(hint) &&
       !canBroadcastBySize(left, conf) && !canBroadcastBySize(right, conf)
   }
 
-  private def probablyHasShuffle(plan: LogicalPlan): Boolean = {
-    plan.collectFirst {
-      case j@Join(left, right, _, _, hint)
-        if isProbablyShuffleJoin(left, right, hint) => j
-      case a: Aggregate => a
-    }.nonEmpty
+  private def probablyHasShuffle(exp: Expression, plan: LogicalPlan): Boolean = {
+    plan.exists {
+      case Join(left, right, _, _, hint)
+        if isProbablyShuffleJoin(left, right, hint) => true
+      case _: Aggregate => true
+      case w: Window
+        if exp.references.subsetOf(AttributeSet(w.partitionSpec.flatMap(_.references))) => true
```

Comment:
How about:
case w: Window => true?

I think `findExpressionAndTrackLineageDown` already checked the pushdown-ability?

Reply:
It is needed because some filters can't push through window, for example:
```sql
SELECT *                                                            
FROM   (SELECT *,                                                   
               Row_number() OVER (PARTITION BY d1 ORDER BY f1) rn   
        FROM   bf1) bf1                                             
       JOIN bf2 ON bf1.c1 = bf2.c2                                  
WHERE  bf2.a2 = 62                                                  
```
```
== Optimized Logical Plan ==
Join Inner, (c1#45922 = c2#45928), Statistics(sizeInBytes=12.3 MiB)
:- Filter (isnotnull(c1#45922) AND might_contain(scalar-subquery#45993 [], xxhash64(c1#45922, 42))), Statistics(sizeInBytes=3.7 KiB)
:  :  +- Aggregate [bloom_filter_agg(xxhash64(c2#45928, 42), 1000000, 8388608, 0, 0) AS bloomFilter#45992], Statistics(sizeInBytes=108.0 B, rowCount=1)
:  :     +- Project [c2#45928], Statistics(sizeInBytes=1278.0 B)
:  :        +- Filter ((isnotnull(a2#45926) AND (a2#45926 = 62)) AND isnotnull(c2#45928)), Statistics(sizeInBytes=3.3 KiB)
:  :           +- Relation default.bf2[a2#45926,b2#45927,c2#45928,d2#45929,e2#45930,f2#45931] parquet, Statistics(sizeInBytes=3.3 KiB)
:  +- Window [row_number() windowspecdefinition(d1#45923, f1#45925 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rn#45976], [d1#45923], [f1#45925 ASC NULLS FIRST], Statistics(sizeInBytes=3.7 KiB)
:     +- Relation default.bf1[a1#45920,b1#45921,c1#45922,d1#45923,e1#45924,f1#45925] parquet, Statistics(sizeInBytes=3.3 KiB)
+- Filter ((isnotnull(a2#45926) AND (a2#45926 = 62)) AND isnotnull(c2#45928)), Statistics(sizeInBytes=3.3 KiB)
   +- Relation default.bf2[a2#45926,b2#45927,c2#45928,d2#45929,e2#45930,f2#45931] parquet, Statistics(sizeInBytes=3.3 KiB)

== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- BroadcastHashJoin [c1#45922], [c2#45928], Inner, BuildRight, false
   :- Filter (isnotnull(c1#45922) AND might_contain(Subquery subquery#45993, [id=#417], xxhash64(c1#45922, 42)))
   :  :  +- Subquery subquery#45993, [id=#417]
   :  :     +- AdaptiveSparkPlan isFinalPlan=false
   :  :        +- ObjectHashAggregate(keys=[], functions=[bloom_filter_agg(xxhash64(c2#45928, 42), 1000000, 8388608, 0, 0)], output=[bloomFilter#45992])
   :  :           +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [id=#415]
   :  :              +- ObjectHashAggregate(keys=[], functions=[partial_bloom_filter_agg(xxhash64(c2#45928, 42), 1000000, 8388608, 0, 0)], output=[buf#45995])
   :  :                 +- Project [c2#45928]
   :  :                    +- Filter ((isnotnull(a2#45926) AND (a2#45926 = 62)) AND isnotnull(c2#45928))
   :  :                       +- FileScan parquet default.bf2[a2#45926,c2#45928] Batched: true, DataFilters: [isnotnull(a2#45926), (a2#45926 = 62), isnotnull(c2#45928)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/spark/hive-3.1.2/connector/kafka-0-10/spark-wareho..., PartitionFilters: [], PushedFilters: [IsNotNull(a2), EqualTo(a2,62), IsNotNull(c2)], ReadSchema: struct<a2:int,c2:int>
   :  +- Window [row_number() windowspecdefinition(d1#45923, f1#45925 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rn#45976], [d1#45923], [f1#45925 ASC NULLS FIRST]
   :     +- Sort [d1#45923 ASC NULLS FIRST, f1#45925 ASC NULLS FIRST], false, 0
   :        +- Exchange hashpartitioning(d1#45923, 5), ENSURE_REQUIREMENTS, [id=#422]
   :           +- FileScan parquet default.bf1[a1#45920,b1#45921,c1#45922,d1#45923,e1#45924,f1#45925] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/spark/hive-3.1.2/connector/kafka-0-10/spark-wareho..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a1:int,b1:int,c1:int,d1:int,e1:int,f1:int>
   +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[2, int, false] as bigint)),false), [id=#427]
      +- Filter ((isnotnull(a2#45926) AND (a2#45926 = 62)) AND isnotnull(c2#45928))
         +- FileScan parquet default.bf2[a2#45926,b2#45927,c2#45928,d2#45929,e2#45930,f2#45931] Batched: true, DataFilters: [isnotnull(a2#45926), (a2#45926 = 62), isnotnull(c2#45928)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/spark/hive-3.1.2/connector/kafka-0-10/spark-wareho..., PartitionFilters: [], PushedFilters: [IsNotNull(a2), EqualTo(a2,62), IsNotNull(c2)], ReadSchema: struct<a2:int,b2:int,c2:int,d2:int,e2:int,f2:int>

```

Reply:
Make sense.  Is it possible to add the logic for Window into `findExpressionAndTrackLineageDown` since it's a dedicated place to check pushdown-ability?

Reply:
+1